From 5a48fd5141a5acfb5135c3340e39670ee4b3b147 Mon Sep 17 00:00:00 2001
From: Arianna Avanzini <avanzini.arianna@gmail.com>
Date: Sat, 26 Jan 2013 01:24:22 +0100
Subject: [PATCH] block, bfq: do not update peak rate if service rate of the
 active queue is zero

It may happen that the amount of service received by the active
queue is so small that the computed service rate is zero. In this
case, do not update the peak rate.
---
 block/bfq-iosched.c |    5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index 5ea128c..c9d57b0 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -1430,9 +1430,6 @@ static int bfq_update_peak_rate(struct bfq_data *bfqd, struct bfq_queue *bfqq,
 	ktime_t delta;
 	int update = 0;
 
-	if (bfqq->entity.service == 0)
-		return 0;
-
 	if (!bfq_bfqq_sync(bfqq) || bfq_bfqq_budget_new(bfqq))
 		return 0;
 
@@ -1473,6 +1470,8 @@ static int bfq_update_peak_rate(struct bfq_data *bfqd, struct bfq_queue *bfqq,
 			 * new_rate = (7/8) * old_rate + (1/8) * bw
 			 */
 			do_div(bw, 8);
+			if (bw == 0)
+				return 0;
 			bfqd->peak_rate *= 7;
 			do_div(bfqd->peak_rate, 8);
 			bfqd->peak_rate += bw;
-- 
1.7.10.4

